hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -1725,6 +1729,201 @@ public void delete(final Path path, final boolean recursive,
     } while (shouldContinue);
   }
 
+  /**
+   * Handles deletion of a path over Blob Endpoint.
+   * @param path path to be deleted
+   * @param recursive defines deletion to be recursive or not
+   * @param tracingContext object to trace the API flows
+   *
+   * @throws IOException exception from server call or if recursive used on
+   * non-empty directory
+   */
+  private void deleteBlobPath(final Path path,
+      final boolean recursive,
+      final TracingContext tracingContext) throws IOException {
+    StringBuilder listSrcBuilder = new StringBuilder();
+    String srcPathStr = path.toUri().getPath();
+    listSrcBuilder.append(srcPathStr);
+    if (!path.isRoot()) {
+      listSrcBuilder.append(FORWARD_SLASH);
+    }
+
+    String listSrc = listSrcBuilder.toString();
+    BlobList blobList = client.getListBlobs(null, listSrc, null, null,
+        tracingContext).getResult().getBlobList();
+    if (blobList.getBlobPropertyList().size() > 0) {
+      LOG.debug(
+          String.format("Path %s has child-blobs", srcPathStr));
+      if (!recursive) {
+        LOG.error(
+            String.format("Non-recursive delete of non-empty directory %s",
+                srcPathStr));
+        throw new IOException(
+            "Non-recursive delete of non-empty directory");
+      }
+      ListBlobQueue queue = new ListBlobQueue(blobList.getBlobPropertyList(),
+          getAbfsConfiguration().getProducerQueueMaxSize(),
+          getAbfsConfiguration().getBlobDirDeleteMaxThread());
+      if (blobList.getNextMarker() != null) {
+        new ListBlobProducer(listSrc, client, queue,
+            blobList.getNextMarker(), tracingContext);
+      } else {
+        queue.complete();
+      }
+      ListBlobConsumer consumer = new ListBlobConsumer(queue);
+      createParentDirectoryIfImplicitOfPathDeleted(path, tracingContext);
+      deleteOnConsumedBlobs(path, consumer, tracingContext);
+      return;
+    }
+    LOG.debug(String.format("Path %s doesn't have child-blobs", srcPathStr));
+    if (!path.isRoot()) {
+      try {
+        getBlobProperty(path, tracingContext);
```

Comment:
As discussed, convert this to a delete call on the path. If its successful, with no exceptions, then trigger a mkdir on the parent path. This will help reduce 4 calls being made to 2 for every delete.

This same change applies to where createParentDirectoryIfImplicitOfPathDeleted is called in the non-empty directory too.

Reply:
Taken this up.
Once delete (with-child path or without) is succesful, we will createDirectory on the parentPath.